### DISCUSSION TITLE
lazy evaluation? benchmarks? updates

### DISCUSSION THREAD

**Discussion by xenoterracide** [marlonrichert/zsh-snap#97] [General]
Does zsnap do any lazy evalution? such as I don't want to do any evaluation until conditions such as prompt available or command pattern matched for completion?

do you have any comparison times with other frameworks?

how does zsnap update plugins and itself? do I need to write a cron/timer?

**Comment by marlonrichert** [comment]
> Does zsnap do any lazy evalution? such as I don't want to do any evaluation until conditions such as prompt available or command pattern matched for completion?

Can you give some concrete examples of how you would want to use this and why?

**Reply by xenoterracide** [reply]
nvm, sdkman, these are really expensive on startup, if I load all my plugins eagerly even with compilation, it'd take like 2 seconds to start.

**Reply by marlonrichert** [reply]
Do you mean these two plugins?
* https://github.com/lukechilds/zsh-nvm
* https://github.com/matthieusb/zsh-sdkman

**Reply by xenoterracide** [reply]
I don't, no, I'm actually referring to the cost of evaluating nvm.sh, etc, itself. I think nvm+sdkman were about 1s combined for evaluation

**Reply by marlonrichert** [reply]
Have you tried using [`znap eval`](https://github.com/marlonrichert/zsh-snap/blob/6ed9cce302f52f579cb8758e6f28e05d5a747c46/.zshrc#L83) for this?

**Reply by xenoterracide** [reply]
nope, is that lazy? or do I have to implement the lazyness

**Reply by marlonrichert** [reply]
Sorry, never mind that. That's not applicable here.

Actually, you're in luck: I forgot that I already implemented this feature some time ago. I just happen to not use it anymore myself. 😅
```zsh
% znap help function
Usage: znap function <func name> ... <init code>
Create lazily loaded functions.
```

Then, the first time when you call any of `<func name> ...`, Znap will run `<init code>` and then execute your command.

So, for `nvm` and `sdkman`, you could add something like this to your `.zshrc` file:
```zsh
znap function nvm 'source $NVM_DIR/nvm.sh'
znap function sdk 'source ~/.sdkman/bin/sdkman-init.sh'
```

This way, `source $NVM_DIR/nvm.sh` and `source ~/.sdkman/bin/sdkman-init.sh` are not executed until you actually try to call `nvm` or `sdk`, respectively.

Try it out and let me know what you think.

**Comment by marlonrichert** [comment]
> do you have any comparison times with other frameworks?

I don’t, but I could make some. What would be meaningful ways to compare  to you?

**Reply by xenoterracide** [reply]
I've been using zinit, but it's occasionally very confusing, and hard to get right. It's considered one of the best performing though. I know there are some other big options, omz is under *new* maintenance, so maybe they've been doing some stuff. I think prezto, zplug, and popular. Maybe this is a good list https://www.slant.co/topics/5827/~zsh-configuration-frameworks. Some people simply benchmark how long the framework adds, I personally think they should benchmark adding and running the same plugins too.  You don't have to add a ton of plugins, but maybe a couple of things of different complexities and note know long it takes to get to the prompt. Frameworks are usually fast enough, it's when you start actually adding the plugins that things get slow, usually because they're immediately executed and don't need to be, and/or they're not compiled.

**Reply by marlonrichert** [reply]
Let's see if I can implement rossmacarthur/zsh-plugin-manager-benchmark#5 at some point.

**Comment by marlonrichert** [comment]
> how does zsnap update plugins and itself? 

`znap status` checks for updates to Znap and of your (plugin) repos in parallel.

`znap pull` without arguments will update Znap and all repos, or you can pass as arguments the specific repos you want to update. 

> do I need to write a cron/timer?

I have not yet implemented an auto-update feature, because I personally have quite a few repos managed by Znap that I don’t want to auto-update. But I am now considering perhaps to implement a selective auto-updatey mechanism.

**Reply by xenoterracide** [reply]
I think, if the service and timer are installed into this directory, they automatically become available to users, users can also install them themselves
```
/usr/lib/systemd/user/powerline-daemon.service is owned by powerline
```

or you could stick it here on install, if not a system package
```
.config/systemd/user/zinit-update.timer
```

I am, was creating these for zinit

```
[Unit]
Description=daily zinit update
Documentation=https://github.com/zdharma/zinit
JobRunningTimeoutSec=60

[Timer]
OnCalendar=daily
AccuracySec=15m
WakeSystem=false
Persistent=true

[Install]
WantedBy=timers.target
```

```
[Unit]
Description=zinit update
Documentation=https://github.com/zdharma/zinit
Requires=network-online.target

[Service]
Type=oneshot
ExecStart=zsh --singlecommand -c 'zinit self-update'
```

and a follow-up service (or it could be associated with a separate timer, if you don't care about the order

```
[Unit]
Description=zinit plugin update
Documentation=https://github.com/zdharma/zinit
Requires=network-online.target
Wants=zinit-update.service
After=zinit-update.service

[Service]
Type=oneshot
ExecStart=zsh --singlecommand -c 'zinit update --all'
```

There's just one problem with this approach, reloading existing shells. I was going to use `zconvey` to do it if I could figure out how to reload `zinit`

People can also use systemd's override to override things if they want.

There is one flaw that i'm working on, there's no way to tell the network is up in a user timer, and I've found out that it's bad for even a system timer.

**Reply by marlonrichert** [reply]
That is all OS-specific. It won’t work, for example, on macOS.

**Reply by xenoterracide** [reply]
Fair

**Reply by xenoterracide** [reply]
At the same time, doing job's is probably easier than some other thing. I don't have a Mac handy anymore, but I suspect it's just a cron job. I'm really not sure about windows, windows is… as per usual probably doing its own thing.

**Reply by marlonrichert** [reply]
#102
